File: 프로그래머스/3/42898. 등굣길/등굣길.java
class Solution {
    
    private static final int mod = 1_000_000_007;
    
    public int solution(int m, int n, int[][] puddles) {
        int[][] DP = new int[n][m];
        
        DP[0][0] = 1;
        
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(isPuddle(i, j, puddles)) {
                    DP[i][j] = 0;
                    continue;
                }
                if(i == 0 && j == 0) continue;
                if(i == 0 && j != 0) {
                    DP[i][j] = DP[i][j - 1];
                    continue;
                }
                if(i != 0 && j == 0) {
                    DP[i][j] = DP[i - 1][j];
                    continue;
                }
                
                DP[i][j] = (DP[i - 1][j] + DP[i][j - 1]) % mod;
            }
        }
        
        return DP[n - 1][m - 1];
    }
    
    private static boolean isPuddle(int x, int y, int[][] puddles) {
        for(int i = 0; i < puddles.length; i++) {
            if(puddles[i][1] == x + 1 && puddles[i][0] == y + 1) return true;
        }
        
        return false;
    }
}

File: 프로그래머스/3/42898. 등굣길/등굣길.js
function solution(m, n, puddles) {
    let D = [];
    
    const mod = 1_000_000_007;
    
    for(let i = 0; i < n; i++) {
        D[i] = [];
        for(let j = 0; j < m; j++) {
            D[i][j] = 0;
        }
    }
    
    D[0][0] = 1;
    
    for(let i = 0; i < n; i++) {
        for(let j = 0; j < m; j++) {
            if(i === 0 && j === 0) continue;
            if(isPuddle(puddles, i, j)) {
                D[i][j] = 0;
            } else {
                if(i === 0) {
                    D[i][j] = D[i][j - 1];
                    continue;
                }
                if(j === 0) {
                    D[i][j] = D[i - 1][j];
                    continue;
                }
                D[i][j] = (D[i - 1][j] + D[i][j - 1]) % mod;
            }
        }
    }
    
    return D[n - 1][m - 1];
}

const isPuddle = (puddles, i, j) => {
    for(let t = 0; t < puddles.length; t++) {
        const x = puddles[t][0] - 1;
        const y = puddles[t][1] - 1;
        
        if(i === y && j === x) {
            return true;
        }
    }
    
    return false;
}